doc/dotNET SDK Proposal.md

```diff
@@ -0,0 +1,232 @@
+
+### Workflow Host Configuration
+
+The user-facing foundation of the SDK is the host. It is configured in a manner that closely follows the conventions for configuring an ASP.NET Core application host ([basic example](../src/Samples/Basic/Temporal.Samples.DesignProcess/Part1_1_TypicalWorkflow.cs#L73-L89) | [more examples](../src/Samples/Basic/Temporal.Samples.DesignProcess/Part4_1_BasicWorkflowUsage.cs#L67-L202) | [another example](../src/Samples/Basic/Temporal.Samples.DesignProcess/Part4_2_BasicWorkflowUsage_MultipleWorkers.cs#L108-L158)).
+
+Users provide workflow implementations and activity implementations in one of two ways:
+ - Pattern based: Entry points of workflow and activity implementations follow specific, well known patterns. Developers use attributes to provide metadata about such entry point to the host. The host automatically loads, and executes workflows and activities. This approach allows writing workflows (and activities) with hardly any scaffolding. 
+- Interface based: Developers implement specific interfaces to create workflows (and activities), and point the host to these implementations. The host automatically loads, and executes workflows and activities. This approach requires a little more scaffolding. However, it also allows additional functionality and helps addressing some advanced scenarios (e.g. dynamic workflow APIs).
+
+### Basic Workflow Execution
+
+Consider a sample workflow of type `SomeWorkflow` that runs a loop executing `SomeActivity` until a certain condition is met. When such condition is eventually met, the workflow is notified via a `NotifyConditionMet` signal.
+
+At application startup, the application is configured to host a worker that can execute workflows (and activities). Then, the worker is configured to poll `SomeTaskQueue` for work items, and to support execution of `SomeWorkflow` instances:
+
+```cs
+public static void Main(string[] args)
+{     
+    IHost appHost = Host.CreateDefaultBuilder(args)
+            .UseTemporalWorkerHost()
+            .ConfigureServices(serviceCollection =>
+            {
+                serviceCollection.AddTemporalWorker()
+                        .Configure(temporalWorkerConfig =>
+                        {
+                            temporalWorkerConfig.TaskQueueMoniker = "SomeTaskQueue";
+                        });
+
+                serviceCollection.AddWorkflowWithAttributes<SomeWorkflow>();
+            })
+            .Build();
+
+    appHost.Run();
+}
+```
+
+We use the `WorkflowAttribute` to mark the class that implements the workflow and to point to the method that implements the main workflow logic. The `WorkflowSignalHandler` is used to specify the method that handles `NotifyConditionMet`-signals:
+
+```cs
+[Workflow(runMethod: nameof(MainAsync))]
```

Comment:
Yes. What would be pros and cons?
I picked this approach to:
- Reduce the number of attributes required.
- Make ensuring that there is exactly one run-method simpler and more intuitive.

But we can certainly go either way on this.

Reply:
Pros:

* No attribute parameter needed on either attribute
* Consistency with query/signal handlers (you could have `queryMethods:` up here too)
* Reduce combining attribute configuration of the workflow method w/ configuration of the workflow (e.g. a workflow's name)
* Clearly marks the entrypoint w/ the attribute (you could even call it `WorkflowEntryPoint` if you wanted)
* Similarity with Java
* Clearly defined at the method level for code readers instead of having to reference the class-level attribute

It's not a huge difference, just clearer to me when you have a workflow entry marked as that type of method, compared to how you're marking other types of special methods.

Reply:
Good point. I would prefer to limit the number of different things as much as possible. If we use an attribute on the main method, then, in my opinion, the class level attribute is superflux. We can use an attribute on the class OR on the main method, but I prefer not both. In that context, the class seems more "natural", but I'd be happe either way.

Reply:
I think marking an entire class as a workflow and marking a method as an entry point is completely reasonable. I think both is just fine. Why is it ok to use both when it's workflow + query but not workflow + entrypoint?

Reply:
A workflow has exactly one entry point. There is always a 1:1 correspondence. Requiring two attributes in 2 places asks the developer to do work that is not necessary, as we already have the information.
Conversely, with handlers, it is 1 workflow to zero-to-many handlers. In some way, each must be marked up.